### PR TITLE
Clean up of next hop HostRecord class.

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

**Comment by rob05c** [comment]
I have very mixed feelings about this.

An ideal solution isn't apparent. Every option has pros and a lot of cons.

>  HostRecord(const HostRecord &) = delete;
>  HostRecord &operator=(const HostRecord &) = delete;

I'm a big fan of deleting the copy and move - HostRecord conceptually should never be copied after initialization.

> struct HostRecord : public ATSConsistentHashNode, public HostRecordCfg {

Not such a big fan of the multiple inheritance. I think the pains of multiple inheritance go without saying.

But the multiple inheritance can easily be avoided by using composition instead. I see the complexity as the bigger question.

Bigger picture, this adds another layer of abstraction, for a very small object. It's an awful lot of complexity for what really ought to be a dead-simple POD class.

Is it really worth all this complexity, the added readability and maintainability costs, to reduce the chance of one small type of bug? I honestly don't know.

Also, this code is potentially leaving `ATSConsistentHashNode.available` uninitialized. Either the `HostRecord` constructor needs to set `available = true;`, or `ATSConsistentHashNode` needs its constructor modified to.

> HostRecord(HostRecordCfg &&o) : HostRecordCfg(o) {}

Is this necessary? I don't think anything besides YAML::Node copies or moves HostRecord, I'd be inclined to delete the move constructor as well.

**Review by jrushford — CHANGES_REQUESTED** [review]

File: proxy/http/remap/NextHopSelectionStrategy.h [hunk 1]
```diff
@@ -103,64 +103,27 @@ struct NHProtocol {
...
-    weight      = o.weight;
-    hash_string = o.hash_string;
-    host_index  = o.host_index;
-    group_index = o.group_index;
-    available   = o.available.load();
-    protocols   = o.protocols;
-    return *this;
-  }
+struct HostRecord : public ATSConsistentHashNode, public HostRecordCfg {
+  std::mutex _mutex;
```
[the remark this answers is not in the document]
Comment (jrushford) [reply proxy/http/remap/NextHopSelectionStrategy.h:114]: oh, I see.  never mind, you added **= delete**

**Comment by ywkaras** [comment]
I guess I was figuring that, since the PR makes the code more than 30 lines shorter, it was making it less complicated.  But maybe I just think too simplistically since I didn't go to graduate school.

File: proxy/http/remap/NextHopSelectionStrategy.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrushford) [reply proxy/http/remap/NextHopSelectionStrategy.h:114]: I think it's fine now, I didn't notice you deleted the operator = and copy constructor.  This make for less maintenance if someone adds a field.

**Comment by rob05c** [comment]
> I guess I was figuring that, since the PR makes the code more than 30 lines shorter, it was making it less complicated

I don't agree that lines of code is a strong indicator of complexity, or difficulty to read and maintain. Thirty lines fewer of assembly would not be less complex, but 30 lines more of Go probably would be.

I do see an additional layer of abstraction, another class, and multiple inheritance as more complex.

But as I said, I have mixed feelings, and I do agree there are advantages. I won't object if @jrushford or anyone else wants to approve.

**Comment by ywkaras** [comment]
It's accurate to observe that I'm using inheritance for something that is not intuitively an is-a relationship.  I could instead have a HostRecordCfg substructure, instead of a base class, but that would mean more changes to the existing code.  But it's fairly common to use inheritance in "tricky" ways in C++:  https://stackoverflow.com/questions/4041447/how-is-stdtuple-implemented

**Review by jrushford — APPROVED** [review]
looks good to me

**Comment by zwoop** [comment]
Cherry-picked to v9.2.x

**Pull request merged**